### ISSUE TITLE
Handwriting and painting

### ISSUE THREAD

**Issue by proninyaroslav** [konradrenner/kolabnotes-android#117]
Hello. I think it would be a good idea. Human writes on the screen and then note is stored as an image.

**Comment by konradrenner** [comment]
Sounds good, do you mean that the image with the handwritten text should be stored inline in the note or as attachment?

The Kolab storage format defines the possibilty to store 0-n attachments to a note.

Do you think it would be useful to attach voice memos to a note?

**Comment by konradrenner** [comment]
Inline images are already supported, but one has to create it with another app

**Comment by proninyaroslav** [comment]
I mean embedded image with handwritten text or drawing. I'm working on the core of editor and if you agree, then I'll start integration into kolab notes.

> Do you think it would be useful to attach voice memos to a note?

Yes, that's a good idea. This can do later.

**Comment by konradrenner** [comment]
Agree :-), thanks for your work!
I will create a new issue for tracking the attachment feature.

**Comment by proninyaroslav** [comment]
While creating text editor I encountered the following problem: all coordinates of the line saved in the ArrayList to implement the ability to undo/redo. But it leads to slower drawing already on hundredth element. Then I decided to set a limit on the 100 elements in the array and in excess of the limit save all the lines in the Bitmap. The only negative - the memory consumption increases by 1.5 - 2 times. Now I stand before a choice: to use or not to use Bitmap. What are your thoughts on this?

**Comment by konradrenner** [comment]
What is the alternative?

**Comment by proninyaroslav** [comment]
I learned a lot of examples of the implementation of image buffering, and came to the conclusion that the Bitmap alternatives at the moment no. In addition, the increase in memory consumption is only once, during the first save buffer. Subsequent savings buffer is almost no effect on consumption.

**Comment by proninyaroslav** [comment]
Experiment with a buffer with a maximum capacity of 10 lines. I drew 30 lines. After every 10 lines of memory consumption is reduced. In the final version, I plan to make a buffer of 100 lines. This will make 100 undo/redo operations, and reduce the number of new Bitmap.

![1](https://cloud.githubusercontent.com/assets/7840559/14013503/4633ffe6-f1c6-11e5-8e84-4920fa988187.png)

**Comment by proninyaroslav** [comment]
> The only negative - the memory consumption increases by 1.5 - 2 times

Fixed. Is left to solve problem with the addition of gesture pinch to zoom and then I'll pull request.

**Comment by proninyaroslav** [comment]
> Is left to solve problem with the addition of gesture pinch to zoom

The main problem is that the gesture pinch to zoom conflicts with multi-finger drawing. When I put two fingers to zoom the canvas, onTouchEvent thinks that I'm drawing two line. Therefore, scaling and drawing line occurs simultaneously.

**Comment by proninyaroslav** [comment]
There are 2 options: 1) not add to pinch to zoom 2) add a special button, activating mode pinch to zoom. What do you think about it?

**Comment by konradrenner** [comment]
I would go with 2 because it does permit pinch to Zoom if one really needs it, but default is pinch to Zoom is disabled

**Comment by konradrenner** [comment]
How did you solve the memory Problem, is it possible that this solution also works for the screen rotation memory Leak?

**Comment by proninyaroslav** [comment]
I don't think. In this case the problem was with the bitmap. To understand the cause of the global memory leaks, it's necessary to examine a dump that I stripped off.

**Comment by konradrenner** [comment]
@proninyaroslav great work so far! I have tested it on smartphone and tablet emulator. Works perfect on smartphone but it does not save the created image in the note on a tablet. Can you please check what the problem is?

I have also added translations and an icon for the "clear canvas"

**Comment by proninyaroslav** [comment]
@konradrenner fixed in 641a698.

**Comment by konradrenner** [comment]
@proninyaroslav great already publishing a new release!
